cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.queue;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.proto.id.ScheduleId;
+
+import java.util.List;
+
+/**
+ * Responsible for keeping track of {@link Job}s, which correspond to schedules that have been triggered,
+ * but not yet executed.
+ */
+interface JobQueue {
+
+  /**
+   * Returns a list of Jobs associated with the given schedule Id.
+   */
+  List<Job> getJobsForSchedule(ScheduleId scheduleId);
+
+  /**
+   * Creates a new Job in the queue or updates an existing Job.
```

Comment:
I meant that if a job already exists, then calling this method updates the existing one in the JobQueue.
Since it is a dataset, yes, calling this `put` method is required to have the update reflected to the actual job in the dataset. This is because you can create the initial job in memory for the first add, so modifications to the Job object aren't reflected by any change in the dataset.

Do you suggest that if the Job was obtained via the dataset's `get`, then it any updates on that Java object should be reflected by updates also to the dataset?